broker.go

```diff
@@ -0,0 +1,584 @@
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+// defaultAddrTTL is the default amount of time that addresses discovered from
+// pubsub messages will remain in the peerstore.  This is twice the default
+// provider poll interval.
+const defaultAddrTTL = 48 * time.Hour
+
+var ErrSourceNotAllowed = errors.New("message source not allowed by policy")
+
+// PeerPolicyHandler is the function signature of a function given to LegBroker
+// that the broker uses to determine whether to allow or reject messages
+// originating from peer passed into the function.  Returning true indicates
+// that messages from the peer are allowed, and false indicated they should be
+// rejected.  Returning an error indicates that there was a problem evaluating
+// the peer, and results in the messages being rejected.
+type PeerPolicyHandler func(peerID peer.ID) (bool, error)
+
+// LegBroker is a message broker for pubsub messages.  LegBroker creates a
+// single pubsub subscriber that receives messages from a gossip pubsub topic,
+// and creates a stateful message handler for each message source peer.  An
+// optional external policy determines whether to allow or deny handling
+// messages from specific peers.
+//
+// Messages to separate peers are handled concurrently, and multiple messages
+// to the same peer are handled serially.  If a handler is busy handling a
+// message, and more messages arrive for the same peer, then the last message
+// replaces the previous unhandled message to avoid having to maintain queues
+// of messages.  Handlers do not have persistent goroutines, but start a new
+// goroutine to handle a single message.
+type LegBroker struct {
+	dtManager   dt.Manager
+	gsExchange  graphsync.GraphExchange
+	tmpDir      string
+	unsubEvents dt.Unsubscribe
+
+	host    host.Host
+	psub    *pubsub.Subscription
+	topic   *pubsub.Topic
+	addrTTL time.Duration
+
+	// dss captures the default selector sequence passed to ExploreRecursiveWithStopNode
+	dss ipld.Node
+
+	handlers      map[peer.ID]*handler
+	handlersMutex sync.Mutex
+	peerPolicy    PeerPolicyHandler
+
+	// Map of CID of in-progress sync to sync done channel.
+	syncDoneChans map[cid.Cid]chan<- struct{}
+	syncDoneMutex sync.Mutex
+
+	// inEvents is used to send a ChangeEvent from a peer handler to the
+	// distributeEvents goroutine.
+	inEvents chan ChangeEvent
+
+	// outEventsChans is a slice of channels, where each channel delivers a
+	// copy of a ChangeEvent to an OnChange reader.
+	outEventsChans []chan ChangeEvent
+	outEventsMutex sync.Mutex
+
+	// cancelBroker cancels broker goroutines
+	cancelBrkr context.CancelFunc
+	// cancelAll cancels pubsub and datatransfer and broker goroutines
+	cancelAll context.CancelFunc
+}
+
+// ChangeEvent notifies an OnChange reader that a specified peer completed a sync.
+type ChangeEvent struct {
+	Cid    cid.Cid
+	PeerID peer.ID
+}
+
+// Defined tells whether the ChangeEvent holds any data.
+func (ce ChangeEvent) Defined() bool {
```

Comment:
Agree, and even that should not occur, except when the event channel closes.In that case the reader should check for channel closed instead of whether the event is defined.